File: transaction.py
"""
This module provides some Cardano blockchain transaction tools
"""
from os import path
from json import loads as json_loads
from subprocess import run as subprocess_run
from tokutils import get_address, get_address_file, get_vkey_file, get_protocol_keydeposit

def split_list(tx):
  """
  split list in two with '+' as a separator
  """
  idx_list = [idx for idx, val in enumerate(tx) if val == '+']
  idx_list1 = [idx+1 for idx, val in enumerate(tx) if val == '+']
  size = len(tx)
  if(size == 2):
    # lovelace only
    return [tx]
  
  # lovelace + tokens
  tx = [tx[i: j] for i, j in
      zip([0] + idx_list1, idx_list + 
      ([size] if idx_list[-1] != size else []))]
  return tx

def build_burn_transaction(network, address, token_name, token_amount, policy_id, utxo, fee, out_file):
  """
  build burn transaction for token
  """
  token_id = policy_id+'.'+token_name
  if token_id not in utxo['balances'].keys():
    print("Cannot burn token", token_name, "with policy", policy_id, ": no token")
    return False
  if utxo['balances'].get(token_id) < token_amount: 
    print("Cannot burn token", token_name, ": not enough token")
    return False
  if policy_id == "":
    print("No policy_id")
    return False
  if utxo['count_utxo'] == 0:
    print("No utxo found")
    return False
  network_era = network['network_era']
  ada_id = 'lovelace'
  asset_id = policy_id+'.'+token_name
  mint = str(-token_amount)+' '+asset_id
  balances = utxo['balances'].copy()
  balances[ada_id] = balances[ada_id]-fee
  balances[asset_id] = balances[asset_id]-token_amount
  if balances[asset_id] < 0 :
    print(address, "does not have", token_amount, token)
    return False
  tx_out = address
  for key, value in balances.items():
    tx_out=tx_out+'+'+str(value)+' '+key

  ttl = calculate_ttl(network)
  run_params = ['cardano-cli', 'transaction', 'build-raw', network_era, '--fee', str(fee)] + utxo['in_utxo'] + \
    ['--ttl', str(ttl), '--tx-out', tx_out, '--mint', mint, '--out-file', out_file]
  print(run_params)
  rc =subprocess_run(run_params, capture_output=False, text=True)
  return rc.returncode == 0

def build_mint_transaction(network, source_address, destination_address, token_name, token_amount, policy_id, utxo, fee, out_file):
  """
  build mint transaction for token
  """
  if token_name is None or token_amount == 0:
    return False
  return build_send_transaction(network, source_address, destination_address, 0, token_name, token_amount, True, policy_id, utxo, fee, out_file)

def build_send_transaction(network, source_address, destination_address, ada_amount, token, token_amount, do_mint, policy_id, utxo, fee, out_file):
  """
  build transfer transaction for token
  """
  if ada_amount < 0 or token_amount < 0:
    print("Can not send negative amount")
    return False
  if ada_amount == 0 and token_amount == 0:
    print("Nothing to send")
    return False
  if policy_id == "" or policy_id == None:
    print("No policy_id")
    return False
  if utxo['count_utxo'] == 0:
    print("No utxo found")
    return False
  if destination_address is None:
    print("No valid destination address")
    return False
  key_deposit = get_protocol_keydeposit(network)
  lovelace_amount = ada_amount * 1000000
  if lovelace_amount == 0:
    lovelace_amount = key_deposit
  elif lovelace_amount < key_deposit:
    if not fee == 0:
      print('To few lovelace for transaction', ada_amount, 'ADA ; you need at least', key_deposit, 'lovelace for transaction')
    return False
  network_era = network['network_era']
  ada_id = 'lovelace'
  balances = utxo['balances'].copy()
  balances[ada_id] = balances[ada_id]-fee-lovelace_amount
  if balances[ada_id] < 0 :
    print(source_address, "does not have", fee+lovelace_amount, "lovelaces (", fee+lovelace_amount/1000000, "ada )")
    return False
  if token is not None:
    asset_id = policy_id+'.'+token
    if not do_mint :
      balances[asset_id] = balances[asset_id]-token_amount
      if balances[asset_id] < 0 :
        print(source_address, "does not have", token_amount, token)
        return False
  tx_out_src=source_address
  for key, value in balances.items():
    tx_out_src=tx_out_src+' +'+str(value)+' '+key
  tx_out_dst = destination_address+'+'+str(lovelace_amount)+' lovelace'
  if token is not None:
    tx_out_dst = tx_out_dst +'+'+str(token_amount)+' '+asset_id
  
  ttl = calculate_ttl(network)
  if do_mint:
    run_params = ['cardano-cli', 'transaction', 'build-raw', network_era, '--fee', str(fee)] + utxo['in_utxo'] + \
     ['--ttl', str(ttl), '--tx-out', tx_out_dst, '--tx-out', tx_out_src, '--mint', str(token_amount)+' '+asset_id, '--out-file', out_file]
  else:
    run_params = ['cardano-cli', 'transaction', 'build-raw', network_era, '--fee', str(fee)] + utxo['in_utxo'] + \
     ['--ttl', str(ttl), '--tx-out', tx_out_dst, '--tx-out', tx_out_src, '--out-file', out_file]
  subprocess_run(run_params, capture_output=False, text=True)
  return True

def calculate_burn_fees(network, address, token, amount, policy_id, utxo, protparams_file):
  """
  calculate fee for burn transaction
  """
  draft_file = get_transaction_file(token, 'draft')
  rc = build_burn_transaction(network, address, token, amount, policy_id, utxo, 0, draft_file)
  if not rc:
    print("Failed to build transaction")
    return None

  rc = subprocess_run(['cardano-cli', 'transaction', 'calculate-min-fee', '--tx-body-file', draft_file, '--tx-in-count', str(utxo['count_utxo']), \
    '--tx-out-count', '1', '--witness-count', '1', '--byron-witness-count', '0', '--protocol-params-file', protparams_file], \
    capture_output=True, text=True)
  min_fee = int(rc.stdout.split(' ')[0])
  return min_fee

def calculate_mint_fees(network, address, token, amount, policy_id, utxo, protparams_file):
  """
  calculate fee for mint transaction
  """
  draft_file = get_transaction_file(token, 'draft')
  rc = build_mint_transaction(network, address, address, token, amount, policy_id, utxo, 0, draft_file)
  if not rc:
    print("Failed to build transaction")
    return None

  rc = subprocess_run(['cardano-cli', 'transaction', 'calculate-min-fee', '--tx-body-file', draft_file, '--tx-in-count', str(utxo['count_utxo']), \
    '--tx-out-count', '1', '--witness-count', '1', '--byron-witness-count', '0', '--protocol-params-file', protparams_file], \
    capture_output=True, text=True)
  min_fee = int(rc.stdout.split(' ')[0])
  return min_fee

def calculate_send_fees(network, source_address, destination_address, ada_amount, token, token_amount, policy_id, utxo, protparams_file):
  """
  calculate fee for transfer transaction
  """
  draft_file = get_transaction_file(token, 'draft')

  rc = build_send_transaction(network, source_address, destination_address, ada_amount, token, token_amount, False, policy_id, utxo, 0, draft_file)
  if not rc: 
    print("Failed to build transaction")
    return None
  rc = subprocess_run(['cardano-cli', 'transaction', 'calculate-min-fee', '--tx-body-file', draft_file, '--tx-in-count', str(utxo['count_utxo']), \
    '--tx-out-count', '1', '--witness-count', '1', '--byron-witness-count', '0', '--protocol-params-file', protparams_file], \
    capture_output=True, text=True)
  min_fee = int(rc.stdout.split(' ')[0])
  return min_fee

def calculate_ttl(network):
  FORWARD_SLOT=300
  run_params = ['cardano-cli', 'query', 'tip', network['network'], str(network['network_magic'])]
  tip = subprocess_run(run_params, capture_output=True, text=True)
  slot = int(json_loads(tip.stdout).get('slotNo'))+FORWARD_SLOT
  return slot

def create_address(network, address_type, addresses_path, address_prefix, name):
  """
  create address based on name
  """
  vkey_file = get_vkey_file(addresses_path, address_prefix, name)
  if not path.exists(vkey_file) :
    print(address_prefix, "verification key missing for", name)
    return None
  addr_file = get_address_file(addresses_path, address_prefix, name)
  if path.exists(addr_file) :
    print(address_prefix, "address already exists for", name)
    return get_address(addr_file)
  network_name = network['network']
  network_magic = str(network['network_magic'])

  run_params = ['cardano-cli', address_type, 'build', network_name, network_magic, \
    '--'+address_prefix+'-verification-key-file', vkey_file, '--out-file', addr_file]
  subprocess_run(run_params, capture_output=False, text=True)
 
  return get_address(addr_file)

def get_transaction_file(token, file_type):
  if file_type == 'draft':
    ext = '.txbody.draft'
  elif file_type == 'ok-fee':
    ext = '.txbody-ok-fee'
  elif file_type == 'sign':
    ext = '.tx.sign'
  else:
    ext = ''

  if token is None:
    file_name = '/tmp/lovelace'+ext
  else:
    file_name = '/tmp/'+token+ext
  return file_name
  
def get_utxo_from_wallet(network, address):
  """
  get utxo from wallet
  """
  if address is None:
    print("address empty")
    return
  network_name = network['network']
  network_magic = str(network['network_magic'])
  network_era = network['network_era']

  env_param = network['env']
  utxo = {}
  tx_disp = subprocess_run(['cardano-cli', 'query', 'utxo', network_name, network_magic, network_era, '--address', address], \
    capture_output=True, text=True, env=env_param)
  tx_list = tx_disp.stdout.split('\n')
  utxo['raw'] = tx_list
  tx_list = [tx.split() for tx in tx_list[2:] if tx != ""]
  t_list = [["--tx-in",tx[0]+'#'+tx[1]] for tx in tx_list]
  # flatten list
  utxo['in_utxo'] = [y for x in t_list for y in x]
  utxo['count_utxo'] = len(tx_list)
  tx_list = [tx[2:] for tx in tx_list]

  tx_list = [split_list(tx) for tx in tx_list]
  # flatten list
  t_list = [y for x in tx_list for y in x]
  utxo['tokens'] = [ (token[1],int(token[0])) for token in t_list]
  return utxo

def sign_mint_transaction(network, skey, policy, ok_fee_file, sign_file):
  """
  sign mint transaction
  """
  network_name = network['network']
  network_magic = str(network['network_magic'])
  rc = subprocess_run(['cardano-cli', 'transaction', 'sign', network_name, network_magic, '--signing-key-file', skey, '--signing-key-file', policy['policy_skey'], \
    '--script-file', policy['policy_script'], '--tx-body-file', ok_fee_file, '--out-file', sign_file], \
    capture_output=False, text=True)
  return

def sign_burn_transaction(network, skey, policy, ok_fee_file, sign_file):
  """
  sign burn transaction
  """
  return sign_mint_transaction(network, skey, policy, ok_fee_file, sign_file)

def sign_send_transaction(network, skey, ok_fee_file, sign_file):
  """
  sign transfer transaction
  """
  network_name = network['network']
  network_magic = str(network['network_magic'])
  rc = subprocess_run(['cardano-cli', 'transaction', 'sign', network_name, network_magic, '--signing-key-file', skey, \
    '--tx-body-file', ok_fee_file, '--out-file', sign_file], \
    capture_output=False, text=True)
  return

def submit_transaction(network, sign_file):
  """
  submit signed transaction on the network
  """
  network_name = network['network']
  network_magic = str(network['network_magic'])
  env_param = network['env']
  rc = subprocess_run(['cardano-cli', 'transaction', 'submit', network_name, network_magic, '--tx-file', sign_file], \
    capture_output=False, text=True, env=env_param)
  return rc

File: Transaction.hs
module Transaction ( buildBurnTransaction, buildMintTransaction, buildSendTransaction, calculateBurnFees, calculateMintFees, createAddress, getTransactionFile,
  FileType(Draft, OkFee, Sign), getUtxoFromWallet, 
  signBurnTransaction, signMintTransaction, submitTransaction, Utxo(Utxo, raw, utxos, nbUtxos, tokens), TransactionType(..) ) where

import System.Directory ( createDirectoryIfMissing, doesFileExist )
import System.IO ( hGetContents)
import System.Process ( createProcess, env, proc, std_out, StdStream(CreatePipe), waitForProcess )
import Data.Maybe ( isNothing, isJust, fromJust, fromMaybe )
import Data.List ( delete, foldl', intercalate )
import Data.List.Split ( splitOn )
import Data.Aeson (decode)
import qualified Data.ByteString.Lazy.Char8 as B
import Tokutils ( Address, AddressType(Payment, Stake), BlockchainNetwork(BlockchainNetwork, network, networkMagic, networkEra, networkEnv),
  calculateTokensBalance, getAddress, getAddressFile, getVkeyFile, getProtocolKeyDeposit, Policy(Policy, policyScript, policyVkey, policySkey, policyId), uglyParse )

data Utxo = Utxo {
    raw :: String
  , utxos :: [String]
  , nbUtxos :: Int
  , tokens :: [(String, Int)]
}

data TransactionType = Burn | Mint | Send  deriving (Eq, Show)


-- utilities ----------------
data FileType = Draft | OkFee | Sign

getTransactionFileExt :: FileType -> String
getTransactionFileExt Draft = ".txbody.draft"
getTransactionFileExt OkFee = ".txbody.ok-fee"
getTransactionFileExt Sign = ".tx.sign"

getTransactionFile :: Maybe String -> FileType -> FilePath 
getTransactionFile Nothing fileType = "/tmp/lovelace" ++ getTransactionFileExt fileType
getTransactionFile (Just token) fileType = "/tmp/" ++ token ++ getTransactionFileExt fileType

-- check if ada and token amount are enough for transaction 
checkSendTransactionAmount :: Int -> Maybe String -> Int -> TransactionType -> String -> Utxo -> Int -> Maybe Int -> IO Bool 
checkSendTransactionAmount adaAmount token tokenAmount transactionType policyId utxo fee keyDeposit
  | adaAmount < 0 || tokenAmount < 0 = do
    putStrLn "Cannot send negative amount"
    return False
  | adaAmount == 0 && tokenAmount == 0 = do
    putStrLn "Nothing to send"
    return False
--  else if isNothing policyId then do
--    putStrLn "No policy id"
--    return False
  | nbUtxos utxo == 0 = do
    putStrLn "No utxo found"
    return False
  | isNothing keyDeposit = do
    putStrLn "No keydeposit value found"
    return False    
  | adaAmount /=0 && adaAmount * 1000000 < fromJust keyDeposit && fee /= 0 = do
    putStrLn $ "To few lovelace for transaction " ++ show adaAmount ++ " ADA ; you need at least " ++ show (fromJust keyDeposit) ++ " lovelace for transaction"
    return False
  | otherwise =
    return True

-- check if ada and token balances are above spending
checkSendTransactionBalance :: [(String, Int)] -> String -> Int -> Int -> Int -> TransactionType -> IO(Bool, [(String, Int)])
checkSendTransactionBalance tokenValues assetId fee lovelaceAmount tokenAmount transactionType = do
  let adaId = "lovelace"
  let balances = calculateTokensBalance tokenValues
  let balances2 = map (\(t, b) -> if t == adaId then (t, b-fee-lovelaceAmount) else (t, b)) balances
  if snd (head (filter(\(t, b) -> t == adaId) balances2)) < 0 then do
    putStrLn $ "The address does not have " ++ show(fee+lovelaceAmount) ++ " lovelaces ( " ++ show(fee+ div lovelaceAmount 1000000) ++ " ada )" 
    return (False, [("",0)])
  else if transactionType /= Mint then do
    let balances3 = map (\(t, b) -> if t == assetId then (t, b-tokenAmount) else (t, b)) balances2
    if snd (head (filter(\(t, b) -> t == assetId) balances3)) < 0 then do
      putStrLn $ "The address does not have " ++ show tokenAmount ++ assetId
      return (False, [("",0)])
    else
      return(True, balances3)
  else
    return(True, balances2)

-- join key value in string
joinkv :: String -> (String,Int) -> String
joinkv acc (key, value) = acc ++ " +" ++ show value ++ " " ++ key

-- build transfer transaction for token
buildTxIn :: [String] -> [String]
buildTxIn utxos = concat ["--tx-in":[u] | u <- utxos]

buildSendTransaction :: BlockchainNetwork -> Address -> Address -> Int -> Maybe String -> Int -> TransactionType -> String -> Utxo -> Int -> FilePath -> IO Bool 
buildSendTransaction bNetwork srcAddress dstAddress adaAmount token tokenAmount transactionType policyId utxo fee outFile = do
  keyDeposit <- getProtocolKeyDeposit bNetwork
  -- need to check isJust keyDeposit
  bool <- checkSendTransactionAmount adaAmount token tokenAmount transactionType policyId utxo fee keyDeposit
  if bool then do
    let adaId = "lovelace"
    let lovelaceAmount = if adaAmount == 0 && (transactionType == Mint || transactionType == Send) then fromJust keyDeposit else adaAmount * 1000000
    let assetId = policyId ++ "." ++ fromMaybe "" token
    (rc, balances) <- checkSendTransactionBalance (tokens utxo) assetId fee lovelaceAmount tokenAmount transactionType
    if rc then do
      let txOutSrc = foldl' joinkv srcAddress (reverse balances)
      let txOutDst = dstAddress ++ "+" ++ show lovelaceAmount ++ " " ++ adaId ++ (if isNothing token then "" else "+" ++ show tokenAmount ++ " " ++ assetId)
      ttl <- calculateTTL bNetwork
      let runParams = ["transaction", "build-raw", networkEra bNetwork, "--fee", show fee] ++ buildTxIn (utxos utxo) ++
            ["--ttl", show ttl, "--tx-out", txOutDst, "--tx-out", txOutSrc] ++ 
            (if transactionType == Mint then ["--mint", show tokenAmount ++" "++assetId] else []) ++ ["--out-file", outFile] 
      (_, Just hout, _, ph) <- createProcess (proc "cardano-cli" runParams){ std_out = CreatePipe }
      r <- waitForProcess ph
      return True
    else
      return False
  else
      return False

-- check burnTransaction
checkBurnTransaction :: Maybe String -> Int -> String -> Utxo -> [(String, Int)] -> IO (Maybe String) 
checkBurnTransaction tokenName tokenAmount policyId utxo balances
  | isNothing tokenName = do
    putStrLn "No token name"
    return Nothing
  | policyId == "" = do
    putStrLn "No policy id"
    return Nothing
  | nbUtxos utxo == 0 = do
    putStrLn "No utxo found"
    return Nothing
  | otherwise = do
    let tokenId = policyId ++ "." ++ fromJust tokenName
    if isNothing $ lookup tokenId balances then do
      putStrLn $ "Cannot burn token " ++ fromJust tokenName ++ " with policy " ++ policyId ++ " : no token"
      return Nothing
    else do
      let tokenBalance = lookup tokenId balances
      if isNothing tokenBalance || fromJust tokenBalance < tokenAmount then do
        putStrLn $ "Cannot burn token " ++ fromJust tokenName ++ " : not enough token"
        return Nothing
      else
        return (Just tokenId)

-- build burn transaction for token
buildBurnTransaction :: BlockchainNetwork -> Address -> Maybe String -> Int -> String -> Utxo -> [(String,Int)] -> Int -> FilePath -> IO Bool 
buildBurnTransaction bNetwork srcAddress token tokenAmount policyId utxo balances fee outFile = do
  assetId <- checkBurnTransaction token tokenAmount policyId utxo balances
  if isJust assetId then do
    keyDeposit <- getProtocolKeyDeposit bNetwork
    let adaId = "lovelace"
    (rc, balances) <- checkSendTransactionBalance (tokens utxo) (fromJust assetId) fee 0 tokenAmount Burn
    if rc then do
      let txOutSrc = foldl' joinkv srcAddress (reverse balances)
      ttl <- calculateTTL bNetwork
      let runParams = ["transaction", "build-raw", networkEra bNetwork, "--fee", show fee] ++ buildTxIn (utxos utxo) ++
            ["--ttl", show ttl, "--tx-out", txOutSrc] ++ 
            ["--mint", show (-tokenAmount) ++" "++fromJust assetId] ++ ["--out-file", outFile]
      (_, Just hout, _, ph) <- createProcess (proc "cardano-cli" runParams){ std_out = CreatePipe }
      r <- waitForProcess ph
      return True
    else
      return False
  else
    return False

-- build mint transaction for token
buildMintTransaction :: BlockchainNetwork -> Address -> Address -> Maybe String -> Int -> String -> Utxo -> Int -> FilePath -> IO Bool 
buildMintTransaction bNetwork srcAddress dstAddress token tokenAmount policyId utxo fee draftFile = do
  if isJust token && tokenAmount /= 0 then
    buildSendTransaction bNetwork srcAddress dstAddress 0 token tokenAmount Mint policyId utxo fee draftFile
  else
    return False

-- calculate fee for burn transaction
calculateBurnFees :: BlockchainNetwork -> Address -> Maybe String -> Int -> String -> Utxo -> [(String,Int)] -> FilePath -> IO (Maybe Int)
calculateBurnFees bNetwork address token amount policyId utxo balances protparamsFile = do
  let draftFile = getTransactionFile token Draft
  bool <- buildBurnTransaction bNetwork address token amount policyId utxo balances 0 draftFile
  if not bool then do
    putStrLn "Failed to build transaction"
    return Nothing
  else do
    let runParams = ["transaction", "calculate-min-fee", "--tx-body-file", draftFile, "--tx-in-count", show(nbUtxos utxo),
          "--tx-out-count", "1", "--witness-count", "1", "--byron-witness-count", "0", "--protocol-params-file", protparamsFile]
    (_, Just hout, _, ph) <- createProcess (proc "cardano-cli" runParams){ std_out = CreatePipe }
    r <- waitForProcess ph
    rc <- hGetContents hout
    let mintFee = read(head (words rc))::Int
    return (Just mintFee)   

-- calculate fee for mint transaction
calculateMintFees :: BlockchainNetwork -> Address -> Maybe String -> Int -> String -> Utxo -> FilePath -> IO (Maybe Int)
calculateMintFees bNetwork address token amount policyId utxo protparamsFile = do
  let draftFile = getTransactionFile token Draft
  bool <- buildMintTransaction bNetwork address address token amount policyId utxo 0 draftFile
  if not bool then do
    putStrLn "Failed to build transaction"
    return Nothing
  else do
    let runParams = ["transaction", "calculate-min-fee", "--tx-body-file", draftFile, "--tx-in-count", show(nbUtxos utxo),
          "--tx-out-count", "1", "--witness-count", "1", "--byron-witness-count", "0", "--protocol-params-file", protparamsFile]
    (_, Just hout, _, ph) <- createProcess (proc "cardano-cli" runParams){ std_out = CreatePipe }
    r <- waitForProcess ph
    rc <- hGetContents hout
    let mintFee = read(head (words rc))::Int
    return (Just mintFee)   

-- calculate network TTL
calculateTTL :: BlockchainNetwork -> IO Int
calculateTTL bNetwork = do
  let forwardSlot=300
  let runParams = ["query", "tip", network bNetwork, show(networkMagic bNetwork)]
  (_, Just hout, _, ph) <- createProcess (proc "cardano-cli" runParams){ std_out = CreatePipe }
  r <- waitForProcess ph
  jsonData <- hGetContents hout
  let slot = forwardSlot + read(uglyParse jsonData "slotNo")::Int 
  return slot

-- create address for owner
createAddress :: BlockchainNetwork -> AddressType -> FilePath -> String -> IO (Maybe Address) 
createAddress bNetwork addressType addressesPath ownerName = do
  let vkFile = getVkeyFile addressesPath addressType ownerName
  boolvk <- doesFileExist vkFile
  if not boolvk then do
    putStrLn $ "verification key missing for " ++ ownerName
    return Nothing
  else do
    let addrFile = getAddressFile addressesPath addressType ownerName
    boolad <- doesFileExist addrFile
    if boolad then do
      putStrLn $ "address already exists for " ++ ownerName
      getAddress addrFile
    else do
      let netName = network bNetwork
      let netMagic = networkMagic bNetwork
      let saddressType = if addressType == Payment then "address" else "stake-address"
      let saddressPrefix = if addressType == Payment then "payment" else "stake"
      let runParams = [saddressType, "build", netName, show netMagic, "--" ++ saddressPrefix ++ "-verification-key-file", vkFile, "--out-file", addrFile]
      (_, Just rc, _, ph) <- createProcess (proc "cardano-cli" runParams){ std_out = CreatePipe }
      r <- waitForProcess ph
      getAddress addrFile

-- get utxo from wallet
getUtxoFromWallet :: BlockchainNetwork -> Address -> IO Utxo
getUtxoFromWallet bNetwork address = do
  let netName = network bNetwork
  let netMagic = networkMagic bNetwork
  let netEra = networkEra bNetwork
  let envParam = Just [("CARDANO_NODE_SOCKET_PATH", networkEnv bNetwork)]
  let runParams = ["query", "utxo", netName, show netMagic, netEra, "--address", address]
  (_, Just hout, _, ph) <- createProcess (proc "cardano-cli" runParams ) { env = envParam } {std_out = CreatePipe }
  r <- waitForProcess ph
  raw <- hGetContents hout

  -- split lines and remove first to lines
  let txList = drop 2  . lines $ raw
  -- get transaction id and index from start of lists and join with #
  let rawUtxos = [ take 2 (words tx) | tx <- txList]
  let utxos = fmap (intercalate "#" ) rawUtxos
-- get tokens from end of lists and build tuples (token, amount)
  let rawTokens = [ drop 2 (words tx) | tx <- txList]
  let ltokens = fmap (filter (/= "+")) rawTokens
  let tokens = concatMap parseTokens ltokens
  return Utxo {raw=raw, utxos=utxos, nbUtxos= length utxos, tokens=tokens}

-- parse transactions list
parseTokens :: [String] -> [(String, Int)]
parseTokens [] = []
parseTokens [x] = []
parseTokens (x:y:xs) = (y,read x::Int):parseTokens xs

-- sign burn transaction
signBurnTransaction :: BlockchainNetwork -> FilePath -> Policy -> FilePath -> FilePath -> IO ()
signBurnTransaction = signMintTransaction

-- sign mint transaction
signMintTransaction :: BlockchainNetwork -> FilePath -> Policy -> FilePath -> FilePath -> IO ()
signMintTransaction bNetwork skeyFile policy okFeeFile signFile = do
  let runParams = ["transaction", "sign", network bNetwork, show(networkMagic bNetwork), "--signing-key-file", skeyFile,
        "--signing-key-file", policySkey policy, "--script-file", policyScript policy, "--tx-body-file", okFeeFile, "--out-file", signFile]
  (_, Just hout, _, ph) <- createProcess (proc "cardano-cli" runParams){ std_out = CreatePipe }
  r <- waitForProcess ph
  return ()

-- submit signed transaction on the network
submitTransaction :: BlockchainNetwork -> FilePath -> IO Bool
submitTransaction bNetwork signFile = do
  let envParam = Just [("CARDANO_NODE_SOCKET_PATH", networkEnv bNetwork)]
  let runParams = ["transaction", "submit", network bNetwork, show(networkMagic bNetwork), "--tx-file", signFile]
  (_, Just hout, _, ph) <- createProcess (proc "cardano-cli" runParams) { env = envParam }{ std_out = CreatePipe }
  r <- waitForProcess ph
  return True
